ckan/model/base.py

```diff
@@ -1,6 +1,231 @@
+            name: str
+            ...
+
+        >>> MyModel.by_name("hello world")
+        None
+
+        >>> MyModel(name="hello world")
+        >>> MyModel.by_name("hello world")
+        <MyModel name=hello world>
+
+        >>> a_obj = MyModel(name="A")
+        >>> d_obj = MyModel(name="D")
+        >>> assert a_obj < d_obj
+    """
+
+    name: str
+
+    def __lt__(self, other: Self) -> bool:
+        return self.name < other.name
+
+    @classmethod
+    def by_name(
+            cls, name: Optional[str], autoflush: bool = True,
+            for_update: bool = False) -> Optional[Self]:
+        """Return first record with the given name.
+        """
+        q = cls.Session.query(cls).autoflush(autoflush).filter_by(name=name)
+        if for_update:
+            q = q.with_for_update()
+
+        return q.first()
+
+
+class TextSearchMixin:
+    """Provide base text-search functionality via LIKE SQL operator.
+
+    Text-search uses columns specified by `text_search_fields` property.
+
+    Example:
+        class MyModel(TextSearchMixin, Base):
+            text_search_fields = ["name]
+            name: str
+            ...
+
+        >>> MyModel(name="hello world")
+        >>> q = model.Session.query(MyModel)
+        >>> MyModel.text_search(q, "llo").all()
+        [<MyModel name=hello world>]
+    """
+    text_search_fields: list[str] = []
+
+    @classmethod
+    def text_search(cls, query: Any, term: str) -> Any:
+        """Add ILIKE filters to the query.
+        """
+        register = cls
+        make_like: Callable[
+            [Any, str], str] = lambda x, y: x.ilike('%' + y + '%')
+        q: Any = sa.null() | sa.null()
+        for field in cls.text_search_fields:
+            attr = getattr(register, field)
+            q = sa.or_(q, make_like(attr, term))
+        return query.filter(q)
+
+
+class ActiveRecordMixin(SessionMixin):
+    """Provides shortcuts for entity lifecycle management.
+    """
+    def save(self) -> None:
+        """Write entity to DB.
+        """
+        self.add()
+        self.commit()
+
+    def add(self) -> None:
+        """Add entity to the session, but don't commit the transaction.
+        """
+        self.Session.add(self)
+
+    def commit_remove(self) -> None:
+        """Commit the transaction and dispose the Session.
+        """
+        self.commit()
+        self.remove()
+
+    def commit(self) -> None:
+        """Commit the transaction.
+        """
+        self.Session.commit()
+
+    def remove(self) -> None:
+        """Dispose the Session.
+        """
+        self.Session.remove()
+
+    def delete(self) -> None:
+        """Mark an instance as deleted.
+        """
+        # stateful objects have this method overridden - see
+        # core.StatefulObjectMixin
+        self.Session.delete(self)
+
+    def purge(self) -> None:
+        """Mark an instance as deleted.
+        """
+        self.Session().autoflush = False
```

Comment:
Why are we setting `autoflush` to False?

Reply:
I assume because the author of these lines didn't want to purge data at that very moment. 11 years passed since this line was added and we have transactions so now it has doesn't look great. 
Still, it won't be very wise to drop such lines inside a PR which only moves code and focuses on preserving existing behavior.